Feature/fractional oct smoothing

f-brinkmann opened the pull request:
### Changes proposed in this pull request:

- Initial draft for fractional octave smoothing

linus-nb committed:
Removed To-Do List

linus-nb committed:
Revert "Removed To-Do List"

This reverts commit bddc79c25bd51bd9e4da3fea111086d411c57af3.

linus-nb committed:
Added To-Do List

linus-nb committed:
Methode integration_limits hinzugefügt.
Konstruktor & Docs angefangen.

linus-nb committed:
Added some docu

linus-nb committed:
Merge branch 'feature-fractional-octave-smoothing-develop' of github.com:linus-nb/pyfar into feature-fractional-octave-smoothing

linus-nb committed:
Merge branch 'develop' of github.com:pyfar-seminar/pyfar into feature-fractional-octave-smoothing-develop

linus-nb committed:
Merge branch 'feature-fractional-octave-smoothing-develop' of github.com:linus-nb/pyfar into feature-fractional-octave-smoothing

linus-nb committed:
Added test and edited calculation of integration limits

linus-nb committed:
Added classmethod and test
Edited gitignore

linus-nb committed:
Correction in calc_integration_limits

linus-nb committed:
Added test for integration limits (not finished)
Bug fixed in calc_integration_limits

linus-nb committed:
removed win type parameter
new test for limit computation
Edited apply() method:
padding of signal by mean,
added dimensions for channels

linus-nb committed:
Correction in exception test

linus-nb committed:
Merge branch 'develop' of github.com:linus-nb/pyfar into develop

linus-nb committed:
Merge branch 'develop' into feature-fractional-octave-smoothing

linus-nb committed:
Merge branch 'develop' into feature-fractional-octave-smoothing-develop

linus-nb committed:
Updated To-Do list
Removed old integration limit test

linus-nb committed:
Removed import in test file

linus-nb committed:
Merge branch 'feature-fractional-octave-smoothing-develop' into feature-fractional-octave-smoothing

linus-nb committed:
Smooth methode for HRTF added.
Test for calc_limits improved

linus-nb committed:
Updated To Do list

linus-nb committed:
edit integration_limits test

linus-nb committed:
Added test for apply method.
Added wrapper function to convert polar to cartesian cmplx. numbers.

linus-nb committed:
Correction in calc_weights test

linus-nb committed:
Added some docs.
Added test for singal smoothing, not completed.

linus-nb committed:
Fixed copy problem.
Added docs.

linus-nb committed:
Merge branch 'feature-fractional-octave-smoothing' into feature-fractional-octave-smoothing-develop

linus-nb committed:
Updated To-Do list

linus-nb committed:
Docs correction

linus-nb committed:
Merge branch 'feature-fractional-octave-smoothing' into develop

linus-nb committed:
Updated Todo List

linus-nb committed:
Removed HRTF method
Changed doc strings to numpy format.

linus-nb committed:
Changed ctor:
no signal data is stored anymore, only number of freq bins.
Signal data is set in the apply() method.
Limits are not stored in the object anymore, only the weights.

linus-nb committed:
Do To list updated.

linus-nb committed:
Merge branch 'feature-fractional-octave-smoothing-develop' into feature-fractional-octave-smoothing

f-brinkmann committed:
Merge branch 'feature/fractional_oct_smoothing' into feature-fractional-octave-smoothing

f-brinkmann committed:
Merge pull request #7 from linus-nb/feature-fractional-octave-smoothing

Merged in semiar pyfar to open draft Pull-Request

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+    return cartesian.reshape(input_shape)
+
+
+def frac_smooth_signal(signal, smoothing_width):
```
I would suggest to
- name this `fract_oct_smooth` (without the signal, because most pyfar functions take signals)
- move this function to `pyfar.dsp.dsp` (pyfar.dsp.fractional_octave_smooting.py should than be imported from pyfar.dsp.dsp)
- There should be an option to apply the smoothing only to the magnitude (abs(spectrum)). Smoothing the complex spectrum will give funky results in most cases. The default should be `true` for this reason

It would also be nice, if the smoothing object could be returned. This could be done if `signal=None` (in analogy to the filter functions)

linus-nb replied:
- Done.
- Done.
- Smoothing is - so far - only performed on the magnitude spectrum of the signal. Applying the weights to the complex spectrum would be interesting if we use complex window functions.

Last: Is on my to do list.

f-brinkmann replied:
Can you double check that smoothing is done on the magnitude values. When I tested it a week ago, the complex spectrum was used. A good way to check this is to smooth a noise signal. If the complex spectrum is used, you will see a high frequency energy loss in the smoothed signal due to the averaging of incoherent data :)

linus-nb replied:
> It would also be nice, if the smoothing object could be returned. This could be done if signal=None (in analogy to the filter functions)

I changed `fract_oct_smooth`method in dsp.py to handle signal or n_bins as input, however this could also be done by the constructor of the class. What is preferred?

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+
+def frac_smooth_signal(signal, smoothing_width):
+    """
+    Method to smooth a given signal.
```
If possible, docstrings should start with a single line, e.g.:
```suggestion
    Smooth spectrum of a signal with fractional octave width.
```

linus-nb replied:
Done.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+    Signal
+        Smoothed Signal
+
+    Raises
```
We usually do not include the `Raises` part. What should be included is a reference to the paper after which this is implemented
```suggestion
    References
    ----------
    .. [1] Name, Title, Journal, ...
```

You can use _[1] inside the docstring to link to the refernece

linus-nb replied:
Done.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+def frac_smooth_signal(signal, smoothing_width):
+    """
+    Method to smooth a given signal.
+    Takes the data of a given signal of shape (n, m), where n is number of
+    channels and m length of the signal.
+    Creates an object of class FractionalSmoothing to compute integration
+    limits and smoothing weights. Finally applies the weights on the input data
+    and returns smoothed signal.
```
- can it handle n-dimensional data? If not you can use `Signal.flatten` before smoothing and `Signal.reshape` after smoothing to make it work.

Generally, the doc string should be more about what is done, than about how it is done. E.g. starting with
`Fractional octave smoothing of a signal spectrum according to _[1]`. I think it would be good to mention how data above fs/2 is treated (i.e. what `data_padder` is doing).

linus-nb replied:
Done.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+    # Compute weights:
+    obj.calc_weights()
+    # Compute smoothed magnitude spectrum
+    data = obj.apply(signal.freq)
```
To make it more pyfar-ish, I would suggest that `apply()` takes and returns a signal. Internally, it uses either `signal.freq` or or `np.abs(signal.freq)`. If the signal is also passed to the class, calling apply without an argument could use the signal stored inside the class. If passing a new signal, we would have to check if `n_bins` and the weights need to be updated.

FFT-Norm:
The FFT-norm of the spectrum should be set to 'none' using `pyfar.fft.normalization()` before the smoothing. The returned signal should have the same FFT norm as the input signal.

linus-nb replied:
`apply()` now takes a signal as argument. Is it necessary to use `signal.copy()` to copy the input signal into a buffer, to avoid changing the signal?
Since I am only changing the domain and fft_norm (both not mutable), I can do `signal_buffer = input_signal`, correct?

f-brinkmann replied:
When I tested it, the input signal seemed not to change. To make sure, can you add a test for this to your to do list. I think we already have code for that on a branch. I will look for it...

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+    # Create smoothing bject
+    obj = FractionalSmoothing(signal.n_bins, smoothing_width)
+    # Compute weights:
+    obj.calc_weights()
```
I did not check this - just posting:
For robustness, the weights should be removed in `FractionalSmoothing` if a related parameter changes (e.g. `n_bins`) and `apply` should either raise an error or calculate the weights itself, if they are missing.

linus-nb replied:
The object can only be constructed with either the whole signal and the smoothing width or the number of frequency bins and the smoothing width. Thereby, the object should contain both parameters once the weights are computed. 
If `n_bins` or `smoothing_width` are changed through the setters, weights are updated.
`n_bins` and `smoothing_width` cannot be deleted.
Added a ValueError if `n_bins` or `smoothing_width` is None in `calc_integration_limits`

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+    if not isinstance(signal, Signal):
+        raise TypeError("Input data must be of type Signal.")
+    # Create smoothing bject
+    obj = FractionalSmoothing(signal.n_bins, smoothing_width)
```
to make it more pyfar-ish, I would pass the entire signal.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+    # Compute smoothed magnitude spectrum
+    data = obj.apply(signal.freq)
+    # Return smoothed signal
+    return Signal(data, signal.sampling_rate, signal.n_samples, domain='freq')
```
We also have to set the `fft_norm` and copy the comment from the input signal.

linus-nb replied:
Done.

f-brinkmann replied:
If you do not have that already. Please add a test to check if the meta-data of the output signal matches that of the input signal. You can take code from `pyfar.signal._assert_matching_metadata`

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        n_bins : int
+            Number of frequency bins of signal.
+        smoothing_width : float, int
+            Width of smoothing window relative to an octave
```
Maybe add 'For example `smoothing_width=1/3` denotes third-octave smoothing' (Also add to the function itself :)

linus-nb replied:
Done.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        smoothing_width : float, int
+            Width of smoothing window relative to an octave
+
+        Raises
```
See comment below - we usually don't include the raised errors :)

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        # Get number of freq bins
+        self._n_bins = n_bins
+        # Save smoothing width:
+        self._smoothing_width = smoothing_width
```
smoothing width could be a property with getter and setter:
```suggestion
        self.smoothing_width = smoothing_width
        
        @property
        def smoothing_width(self):
            return self._smoothing_width
        
        @smoothing_width.setter
        def smoothing_width(self, value):
            self.smoothing_width = value
            # The setter could also reset the weights if the smoothing width changed
```

linus-nb replied:
Done. n_bins and smoothing_width are now properties, the weights matrix is updated in the setter method.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+    def calc_integration_limits(self):
+        """
+        Computes integration limits for each frequency bin
+        k. For each k two arrays are computed, one containing the upper, one
```
Start with single line doc-str:
```suggestion
        Compute integration limits for each frequency bin.
        
        For each k two arrays are computed, one containing the upper, one
```

You could also reference the equation number already in the docstring.

linus-nb replied:
Done. Also added equation references.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        # Remove padded samples:
+        magnitude = magnitude[:, :self._n_bins]
+        # Copy phase from original data
+        phase = np.angle(data)
```
using the original phase is dangerous because it can cause an invalid spectrum. On possibility might be to have a parameter `phase_type` to let the user chose. A good default would be `'zero'`, i.e., `phase=0`. The current way could be `'original'`. If we have methods for minimum and linear phase, we could add that later.

linus-nb replied:
Added phase_type as property with valid phase types: original, zero, minimal, linear.

f-brinkmann replied:
Just a convention - `minmum phase` if more common than `minimal phase` :)

linus-nb replied:
I never named it  `minimal` , must have been a confusion in my comment ;)

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        # Copy phase from original data
+        phase = np.angle(data)
+        # Return array in cartesian form:
+        return polar2cartesian(magnitude, phase)
```
Does this do anything else than `complex_spectrum = magnitude * np.exp(1j * phase)`? If no, we could use this instead of the rather lengthy expression and move `polar2cartesian` to `pyfar.dsp`.

linus-nb replied:
It is wrapper for cmath.rect method applied on two ndarrays of same shape:
```   
if not (amplitude.shape == phase.shape): 
        raise ValueError("Arrays must have same shapes.")
# Save input shape
input_shape = amplitude.shape

# Reshape for cmath rect to 1D arrays:
reshaped_amplitude = amplitude.reshape(1, -1)[0]
reshaped_phase = phase.reshape(1, -1)[0]

# Apply cmath rect on elements:
cartesian = np.array([cm.rect(a, p) for a, p in zip(reshaped_amplitude, reshaped_phase)])

return cartesian.reshape(input_shape)
```

f-brinkmann replied:
I see. I would suggest to use `complex_spectrum = magnitude * np.exp(1j * phase)` instead. This is less code, does not require reshaping, and does not have the loop of the list comprehension.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        return polar2cartesian(magnitude, phase)
+
+
+def lim_padder(low, up, size):
```
I think this should be a class method of `FractionalSmoothing`. Or is this useful outside of this class?

linus-nb replied:
I changed it to static method since it is just a small helper function for computing the integration limits.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        lower_limit.reshape(1, -1)))
+
+
+def data_padder(data, pad_width, mean_size):
```
I think this should be a class method of `FractionalSmoothing`. Or is this useful outside of this class?

linus-nb replied:
Also changed this one to static method. It only gets called in the apply() method once.

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,317 @@
+        # Save smoothing width:
+        self._smoothing_width = smoothing_width
+
+    def calc_integration_limits(self):
```
Does this calculate the integration limits, or the values of the integral at each integration limit / bin? I did not fully understand the code (the results seems to be correct, though)

linus-nb committed:
Merge branch 'develop' of github.com:pyfar-seminar/pyfar into develop

linus-nb committed:
Limits/weights changed to sparse matrix
Todo: fix tests

linus-nb committed:
Fixed tests

linus-nb committed:
Added test for weight calc at specified freq

linus-nb committed:
Updated Todo list

linus-nb committed:
- Changed apply input argument to signal
- Changes lim_padder and data_padder to static methods
- added properties: n_bins, smoothing_width, phase_type
 (default is 'Zero'
- Input signal can now be n-dimensional
- moved fract_oct_smooth to dsp.dsp

linus-nb committed:
Fixed setter and added phase_type as input parameter in fract_oct_smooth

f-brinkmann commented on pyfar/dsp/TODOs.md:
```diff
@@ -0,0 +1,41 @@
+* [ ] apply() fixen
+* [ ] Get- & Set-Methoden testen
+* [ ] Smoothing eines mehrkanaliges Eingangssignal(HRTF) testen
+* [ ] Alle Methoden getestet?
```
You can get a coverage report from pytest if you run it from the command line:
`pytest --cov=. --cov-report=html`

Thats a good to see if all relevant parts are tested :)

f-brinkmann commented on pyfar/dsp/dsp.py:
```diff
@@ -187,3 +188,49 @@ def spectrogram(signal, dB=True, log_prefix=20, log_reference=1,
+    Smooth magnitude spectrum of a signal with fractional octave width
+    according to _[1].
+
+    Takes the data of a given signal of shape (n, m), where n is number of
```
The method should work for any signal - `(n, m)` here is this not necessary

f-brinkmann commented on pyfar/dsp/dsp.py:
```diff
@@ -187,3 +188,49 @@ def spectrogram(signal, dB=True, log_prefix=20, log_reference=1,
+
+    References
+    ----------
+    .. [1] JO. G.. Tylka, BR. B.. Boren, and ED. Y.. Choueiri,
```
```suggestion
    .. [1] J. G. Tylka, B. B. Boren, and E. Y. Choueiri,
```

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,423 @@
+
+        References
+        ----------
+        .. [1] JO. G.. Tylka, BR. B.. Boren, and ED. Y.. Choueiri,
```
see above - maybe get rid of the double initials and dots...

f-brinkmann commented on pyfar/dsp/fractional_octave_smoothing.py:
```diff
@@ -0,0 +1,423 @@
+        Compute integration limits for each frequency bin as in eq. (17) in
+        _[1].
+
+        For each k two arrays are computed, one containing the upper, one
```
Maybe better: 'For each frequency bin k ...'

letreoj committed:
Add files via upload

letreoj committed:
Create pyfar_test_signals.py

letreoj committed:
Rename dirac_generator.py to pyfar_test_signals/dirac_generator.py

letreoj committed:
Rename noise_generator.py to pyfar_test_signals/noise_generator.py

letreoj committed:
Rename sine_generator.py to pyfar_test_signals/sine_generator.py

letreoj committed:
Rename sweep_generator.py to pyfar_test_signals/sweep_generator.py

letreoj committed:
Delete pyfar_test_signals.py

linus-nb committed:
Fixed apply() test,
Updated To Do list,
Removed polar2cartesian,
Added flag to calc weights,
fract_oct_smooth method now takes signal or n_bins as input.

linus-nb committed:
Updated To Do List
Added phase handling test

linus-nb committed:
Fixed phase handling

linus-nb committed:
Merge branch 'feature/test_signals' into feature/fractional_oct_smoothing

linus-nb committed:
Added setters tests,
fixed setters,
changed fft_norm of return signal in apply() to input fft norm

linus-nb committed:
Merge branch 'develop' into feature/fractional_oct_smoothing

f-brinkmann committed:
Revert "Merge branch 'feature/test_signals' into feature/fractional_oct_smoothing"

This reverts commit 740b2d1bee6a1f9899eb0fc421502e0a23cf044d, reversing
changes made to 380d4c37079013be43b0474eda75f94f0b2e8379.

linus-nb committed:
added new calc_weigths method, test update necessary.

linus-nb committed:
Added new weight calc method.
Added new smoothing method by loop.
Added Enum classes for phase and padding type.
Added test for loop method.

linus-nb committed:
Edit data_padder to handle padding types.

linus-nb committed:
Fixed bug in dsp.fract_oct_smooth,
changed apply method to new version with loop,
tested new apply method

linus-nb committed:
Merge remote-tracking branch 'origin/develop' into feature/fractional_oct_smoothing

linus-nb committed:
Added notebook for presenation.
Updated dsp.fract_oct_smooth.
Added mean padding to new apply method, added padding_type setter.
Added new fixtures to tests,
added test for data_padder, padding handling and smoothing.

linus-nb committed:
Added some exception tests.

linus-nb committed:
Some renaming, moved array generation out of loop in apply for speed up,
added presentation script

linus-nb committed:
Edit smoothing test,
added multiprocessing benchmark script,
edit in presentation script

linus-nb committed:
Added c_loop version
